src/wallet/rpcdisclosure.cpp

```diff
@@ -0,0 +1,296 @@
+    }
+    o.push_back(Pair("jsIndex", pd.payload.js));
+
+    if (pd.payload.n < 0 || pd.payload.n > 1) {
+        errs.push_back("Payment disclosure refers to an invalid output index");
+    }
+    o.push_back(Pair("outputIndex", pd.payload.n));
+
+    if (pd.payload.version != 0) {
+        errs.push_back("Payment disclosure refers to an unknown version");
+    }
+    o.push_back(Pair("version", pd.payload.version));
+
+    o.push_back(Pair("ephemeralSymmetricKey", pd.payload.esk.ToString()));
+    o.push_back(Pair("message", pd.payload.message));
+    o.push_back(Pair("joinSplitPubKey", tx.joinSplitPubKey.ToString()));
+
+    // Verify the payment disclosure was signed using the same key as the transaction i.e. the joinSplitPrivKey.
+    uint256 dataToBeSigned = SerializeHash(pd.payload, SER_GETHASH, 0);
+    bool verified = (crypto_sign_verify_detached(pd.payloadSig.data(),
+        dataToBeSigned.begin(), 32,
+        tx.joinSplitPubKey.begin()) == 0);
+    o.push_back(Pair("signatureVerified", verified));
+    if (!verified) {
+        errs.push_back("Payment disclosure signature does not match transaction signature");        
+    }
+   
+    // Check the payment address is valid
+    PaymentAddress zaddr = pd.payload.zaddr;
+    CZCPaymentAddress address;
+    if (!address.Set(zaddr)) {
+        errs.push_back("Payment disclosure refers to an invalid payment address");        
+    } else {
+        o.push_back(Pair("paymentAddress", address.ToString()));
+
+        try {
+            // Decrypt the note to get value and memo field
+            JSDescription jsdesc = tx.vjoinsplit[pd.payload.js];
+            uint256 h_sig = jsdesc.h_sig(*pzcashParams, tx.joinSplitPubKey);
+
+            ZCPaymentDisclosureNoteDecryption decrypter;
+
+            ZCNoteEncryption::Ciphertext ciphertext = jsdesc.ciphertexts[pd.payload.n];
+
+            uint256 pk_enc = zaddr.pk_enc;
+            auto plaintext = decrypter.decryptWithEsk(ciphertext, pk_enc, pd.payload.esk, h_sig, pd.payload.n);
+
+            CDataStream ssPlain(SER_NETWORK, PROTOCOL_VERSION);
+            ssPlain << plaintext;
+            NotePlaintext npt;
+            ssPlain >> npt;
+
+            string memoHexString = HexStr(npt.memo.data(), npt.memo.data() + npt.memo.size());
+            o.push_back(Pair("memo", memoHexString));
+            o.push_back(Pair("value", ValueFromAmount(npt.value)));
+            
+            // Check the blockchain commitment matches decrypted note commitment
+            uint256 cm_blockchain =  jsdesc.commitments[pd.payload.n];
+            Note note = npt.note(zaddr);
+            uint256 cm_decrypted = note.cm();
+            bool cm_match = (cm_decrypted == cm_blockchain);
+            o.push_back(Pair("commitmentMatch", cm_match));
+            if (!cm_match) {
+                errs.push_back("Payment disclosure derived commitment does not match blockchain commitment");
```

Comment:
`Commitment derived from payment disclosure does not match blockchain commitment`

Also, `blockchain` or `block chain`?

Reply:
Ok. Always blockchain :-)